src/compiler/utilities.ts

```diff
@@ -2713,6 +2713,13 @@ namespace ts {
         return forEach(supportedTypeScriptExtensions, extension => fileExtensionIs(fileName, extension));
     }
 
+    /** Return ".ts" or ".tsx" if that is the extension. */
+    export function tryExtractTypeScriptExtension(fileName: string): string | undefined {
+        return tryFind(supportedTypescriptExtensionsWithDtsFirst, extension => fileExtensionIs(fileName, extension));
+    }
+    // Must have '.d.ts' first because if '.ts' goes first, that will be detected as the extension instead of '.d.ts'.
+    const supportedTypescriptExtensionsWithDtsFirst = supportedTypeScriptExtensions.slice().reverse();

```

Comment:
Seems like something you probably want to make explicit in case anyone relies on that behavior.